streaming/base/shared/prefix.py

```diff
@@ -54,48 +62,101 @@ def _unpack_locals(data: bytes) -> Set[str]:
+
+    Local leader walks the existing shm prefixes starting from zero, verifying that there is no
+    local working directory overlap. When attaching to an existing shm fails, we have reached the
+    end of the existing shms. We will register the next one.
+
+    Args:
+        my_locals_set (Set[str]): Our local working directories.
+
+    Returns:
+        int: Next available prefix int.
+    """
+    prefix_int = 0
+    for prefix_int in _each_prefix_int():
+        prefix = f'{prefix_int:06}'
+        name = f'{prefix}_locals'
+        try:
+            shm = SharedMemory(name, False)
+        except FileNotFoundError:
+            break
+        their_locals_set = _unpack_locals(bytes(shm.buf))
+        both = my_locals_set & their_locals_set
+        if both:
+            raise ValueError(f'Reused local directory: {sorted(my_locals_set)} vs ' +
+                             f'{sorted(their_locals_set)}. Provide a different one.')
+    return prefix_int
+
+
+def _check_and_find_retrying(my_locals_set: Set[str], retry: int) -> int:
+    """Find the next available prefix while checking existing dirs for overlap.
+
+    If an overlap is found, sleeps for a tick and then tries again, up to "retry" times. We allow
+    this grace period because modifying python shared memory in a destructor intermediated through
+    a numpy array appears to be racy.
+
+    Args:
+        my_locals_set (Set[str]): Our local working directories.
+        retry (int): Number of retries upon failure before raising an exception.
+
+    Returns:
+        int: Next available prefix int.
+    """
+    if retry < 0:
+        raise ValueError(f'Specify at least zero retries (provided {retry}).')
+    errs = []
+    for _ in range(1 + retry):
+        try:
+            return _check_and_find(my_locals_set)
+        except ValueError as err:
+            errs.append(err)
+            sleep(TICK)
+    raise errs[-1]
+
+
+def get_shm_prefix(my_locals: List[str], world: World, retry: int = 7) -> Tuple[str, SharedMemory]:
+    """Register or lookup our shared memory prefix.
+
+    Args:
+        my_locals (List[str]): Local working dir of each stream, relative to /. We need to verify
+            that there is no overlap with any other currently running StreamingDataset.
+        world (World): Information about nodes, ranks, and workers.
+        retry (int): Number of retries upon failure before raising an excpetion. Defaults to ``7``.
+
+    Returns:
+        Tuple[str, SharedMemory]: Shared memory prefix and object. The name is required to be very
+            short due to limitations of Python on Mac OSX.
+    """
+    # Check my locals for overlap.
+    my_locals_set = _check_self(my_locals)
+
+    # First, the local leader registers the first available shm prefix, recording its locals.
+    if world.is_local_leader:
+        prefix_int = _check_and_find_retrying(my_locals_set, retry)
```

Comment:
If `_check_and_find_retrying` errors, this return value will be the error message. Is that intended? Should there be some error handling here?

Reply:
eh?